src/Xamarin.Android.Build.Tasks/MSBuild/Xamarin/Android/Microsoft.Android.Sdk.Bindings.Gradle.targets

```diff
@@ -0,0 +1,114 @@
+          Condition= " Exists('%(AndroidGradleProjectReference.FullPath)') "/>
+      <_GradleInputs Include="@(_AllGradleFiles)"
+          Condition="  '%(Extension)' == '.java'
+                    or '%(Extension)' == '.kt'
+                    or '%(Extension)' == '.kts'
+                    or '%(Extension)' == '.gradle'
+                    or '%(Extension)' == '.properties'
+                    or '%(Extension)' == '.xml' " />
+      <_GradleInputs Include="$(MSBuildProjectFullPath)" />
+      <_GradleInputs Include="@(_AndroidMSBuildAllProjects)" />
+    </ItemGroup>
+  </Target>
+
+  <Target Name="BuildAndroidGradleProjects"
+      Condition=" '@(AndroidGradleProjectReference->Count())' != '0' "
+      DependsOnTargets="_SetupMSBuildAllProjects;AndroidPrepareForBuild;_GetBuildAndroidGradleProjectsInputs"
+      Inputs="@(_GradleInputs)"
+      Outputs="$(_BuildAndroidGradleProjectsStamp)" >
+    <!-- Delete outputs folder if inputs have changed to force a partial rebuild -->
+    <RemoveDir Directories="@(AndroidGradleProjectReference->'%(OutputPath)outputs')" />
+
+    <!-- Run assemble task for project outputs, android app and library project outputs are currently supported -->
+    <Gradle ToolPath="%(AndroidGradleProjectReference.FullPath)"
+        Command="assemble%(AndroidGradleProjectReference.Configuration)"
+        ModuleName="%(AndroidGradleProjectReference.ModuleName)"
+        OutputPath="%(AndroidGradleProjectReference.OutputPath)"
+        IntermediateOutputPath="$(MSBuildProjectDirectory)/$(IntermediateOutputPath)"
+        AndroidSdkDirectory="$(AndroidSdkDirectory)"
+        JavaSdkDirectory="$(JavaSdkDirectory)" >
+      <Output TaskParameter="AppOutputs" ItemName="_AndroidGradleProjectAppOutputs" />
+      <Output TaskParameter="LibraryOutputs" ItemName="_AndroidGradleProjectLibraryOutputs" />
+    </Gradle>
+
+    <WriteLinesToFile
+        File="%(AndroidGradleProjectReference.OutputPath)libraryoutputs.txt"
+        Lines="@(_AndroidGradleProjectLibraryOutputs)"
+        Overwrite="true"
+        WriteOnlyWhenDifferent="true"
+        Condition=" '@(_AndroidGradleProjectLibraryOutputs->Count())' != '0' "
+    />
+
+    <Touch Files="$(_BuildAndroidGradleProjectsStamp)" AlwaysCreate="True" />
+    <ItemGroup>
+      <FileWrites Include="$(_BuildAndroidGradleProjectsStamp);%(AndroidGradleProjectReference.OutputPath)**/*" />
+      <!-- Copy app output files to $(OutputPath) -->
+      <None Include="@(_AndroidGradleProjectAppOutputs)" CopyToOutputDirectory="PreserveNewest" Link="%(Filename)%(Extension)" />
+    </ItemGroup>
+  </Target>
+
+  <!-- Automatically add library outputs to be bound or otherwise processed later in the build
+        This is done in a separate target to ensure @(AndroidLibrary) items are added even when BuildAndroidGradleProjects is skipped -->
+  <Target Name="_IncludeAndroidGradleProjectsOutputs"
```

Comment:
@dellis1972 @jonathanpeppers I'm struggling to find the best way to _always_ add the outputs of the `Gradle` task to `AndroidLibrary`, even when `BuildAndroidGradleProjects` is skipped entirely.

The current approach to write library outputs to a file and read it in a separate target seems to work well for incremental build scenarios, but it doesn't feel like the most sane or cleanest option...

Reply:
You can hash the outputs into a file then use the file as the input? That way if the outputs change so will then hash. 
We've done this in a few places. See https://github.com/dotnet/android/blob/ac8a922d495dff28fdc29310d509a21fa139b883/src/Xamarin.Android.Build.Tasks/Microsoft.Android.Sdk/targets/Microsoft.Android.Sdk.AndroidLibraries.targets#L54

Reply:
I'll need to retain the path of the file so that it can be added to the corresponding `AndroidLibrary` item that is being created, so I am not sure if hashing the itemgroup will help me in this case unfortunately.

Reply:
ok I missread the issue. 

If a Target is Skipped the `PRopertyGroup` and `ItemGroup` within the target will still be evaluated. So if you know where to look you can populate an ItemGroup even if the target does not run.

Reply:
Hmm so I should be able to avoid this extra target if I no longer have the `Gradle` task identifying the project build outputs, and instead add that lookup to the targets?

Reply:
One way I've done this before is have the Task output the items, but write that list to a file which can be read if the Outputs ItemGroup is empty. I can't remember where I did that though. I seem to remember this coming up in Maui.

Reply:
Thanks! Looks like moving output collection to the target instead of the task works well here.